src/utils/amp.cu

```diff
@@ -0,0 +1,148 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2023, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "lbann/utils/amp.hpp"
+#include "lbann/utils/profiling.hpp"
+#include "lbann/utils/gpu/helpers.hpp"
+
+namespace lbann {
+namespace amp {
+
+namespace {
+
+// std::isfinite does not work for half precision, so we handle that here.
+// There does not seem to be a CUDA/ROCm version of isfinite for __half, so
+// we handle it by converting to float.
+
+template <typename T>
+__host__ __device__ __forceinline__ bool amp_is_finite(T val) {
+  return std::isfinite(val);
+}
+
+#ifdef LBANN_HAS_GPU_FP16
+template <>
+__host__ __device__ __forceinline__ bool amp_is_finite<fp16>(fp16 val) {
+  return std::isfinite(static_cast<float>(val));
+}
+#endif
```

Comment:
```suggestion
template <typename T>
__host__ __device__ __forceinline__ bool amp_is_finite(T val) {
  return lbann::gpu_lib::isfinite(val);
}
```
Or just use this function throughout. Don't forget to `#include <lbann/utils/gpu/gpu_lib.hpp>`.

Reply:
Done. Also _do_ forget to `#include "lbann/utils/gpu/gpu_lib.hpp"`, apparently that header is not meant to be included directly.